rfcs/hse/0004-kvdb-kvs-introspection-apis.md

```diff
@@ -0,0 +1,160 @@
+# KVDB/KVS Introspection APIs
+
+## Problem
+
+HSE does not currently have any way to introspect KVDB/KVS information.
+
+## Requirements
+
+HSE should provide APIs for the following:
+
+- KVDB
+  - create parameters
+  - runtime parameters
+  - home
+  - KVS names (done)
+- KVS
+  - create parameters
+  - runtime parameters
+  - name
+
+Any REST APIs will support the notion of multiple KVDBs.
+
+## Non-Requirements
+
+- Will not add support for multiple KVDBs in other areas of HSE
+
+## Solution
+
+Create new APIs in C, Python, and REST in order to expose this information. REST
+APIs are necessary because you cannot introspect a KVDB open in another process.
+
+### Overview
+
+#### C APIs
+
+The following C APIs will be added behind the experimental wall:
+
+```c
+/* Should this API return an error in the case of a NULL KVDB?
+ *
+ * Returns a pointer to the kvdb_home member in ikvdb struct.
+ */
+const char *
+hse_kvdb_home_get(const struct hse_kvdb *kvdb);
+
+/* Should this API return an error in the case of a NULL KVDB?
+ *
+ * Returns a pointer to the name member in ikvs struct.
+ */
+const char *
+hse_kvs_name_get(const struct hse_kvs *kvs);
+
+/* Returns a JSON representation of the parameter value, ie the value in
+ * key=value.
+ */
+hse_err_t
+hse_kvdb_param_get(const struct hse_kvdb *kvdb, const char *param, char *buf, size_t buf_len);
+
+/* Returns a JSON representation of the parameter value, ie the value in
+ * key=value.
+ */
+hse_err_t
+hse_kvs_param_get(const struct hse_kvs *kvs, const char *param, char *buf, size_t buf_len);
+
+enum hse_mclass {
+  HSE_MCLASS_CAPACITY,
+  HSE_MCLASS_STAGING,
+};
+
+struct hse_mclass_storage_info {
+  uint64_t total_bytes;     /**< total space in the file-system containing this kvdb */
+  uint64_t available_bytes; /**< available space in the file-system containing this kvdb */
+  uint64_t allocated_bytes; /**< allocated storage space for a kvdb */
+  uint64_t used_bytes;      /**< used storage space for a kvdb */
+  char     path[PATH_MAX];  /**< path to media class */
+};
+
+/* Populates the above storage info struct with given information.
+ *
+ * Con: Since we only track data owned by each media class, we drop the storage
+ * statistics about the kvdb.meta file, but I don't see that as being super
+ * important.
+ */
+hse_err_t
+hse_kvdb_storage_info_get(const struct hse_kvdb *kvdb, enum hse_mclass mclass, struct hse_mclass_storage_info *info);
```

Comment:
That would make sense to me!

Reply:
Although it feels like the CLI commands should be updated then.

```
hse mclass add
hse mclass info
```

Reply:
or maybe just `hse storage info` has to change to `hse mclass info`. Too many decisions.

Reply:
Maybe it makes sense just to keep the API and cli as storage info. Less churn, and it still sounds like a good name to me.

Reply:
The CLI commands need not be updated.

"hse storage info" provides aggregated media class stats. We can output per-media class stats in the verbose mode.

"hse storage add" can be used to add multiple media classes in a single CLI command.

"hse storage profile" takes a FS path today. If/when we update it to take KVDB home as input, it can pick one or more media classes to profile.

The point being the "hse storage" command can use multiple media classes at the same time, so that can remain unchanged.

Reply:
Oh good point!